src/snowflake/snowpark/_internal/udf_utils.py

```diff
@@ -825,142 +946,180 @@ def resolve_imports_and_packages(
     skip_upload_on_content_match: bool = False,
     is_permanent: bool = False,
     force_inline_code: bool = False,
-) -> Tuple[str, str, str, str, str, bool]:
-    import_only_stage = (
-        unwrap_stage_location_single_quote(stage_location)
-        if stage_location
-        else session.get_session_stage(statement_params=statement_params)
-    )
-
-    upload_and_import_stage = (
-        import_only_stage
-        if is_permanent
-        else session.get_session_stage(statement_params=statement_params)
-    )
+) -> Tuple[
+    Optional[str],
```

Comment:
`Optional` should be used very cautiously here.

Reply:
Discussed the override mechanism here, though that can be a fast-follow to this PR.

Reply:
Please add a TODO comment with the JIRA number/Github issue link here.